gms/gossiper.cc

```diff
@@ -861,9 +863,14 @@ future<> gossiper::failure_detector_loop() {
     logger.info("failure_detector_loop: Finished main loop");
 }
 
-future<> gossiper::replicate_live_endpoints_on_change() {
+// replicate shard 0 endpoints to other shards
+// if live_members_out is not nullptr, it is updated with the current live members
+future<> gossiper::replicate_live_endpoints_on_change(std::set<inet_address>* live_members_out) {
     assert(this_shard_id() == 0);
     auto lock = co_await get_units(_endpoint_update_semaphore, 1);
+    if (live_members_out != nullptr) {
+        *live_members_out = get_live_members();
+    }
```

Comment:
I'd either return that as the future value, or refactor the code to have `with_endpoint_update_semaphore(std::function<future<>(gossiper&))` that does `container().invoke_on(0` and acquires the lock there, and run the functor on the shard 0 instance, and extract the rest of this function to `replicate_live_endpoints_on_change_locked`

making this function look something like:

```c++
with_endpoint_update_semaphore([this] (gossiper& g) {
    return g.replicate_live_endpoints_on_change_locked();
});
```

while `get_live_members_synchronized` can do something like:
```c++
with_endpoint_update_semaphore([this] (gossiper& g) -> future<std::set<inet_address>> {
    auto ret = get_live_members();
    co_await g.replicate_live_endpoints_on_change_locked();
    co_return ret;
});
```

Reply:
I think
```cpp
{
  auto lock = co_await get_units(_endpoint_update_semaphore, 1);
  ... rest of the code ...
}
```
is simpler than
```cpp
with_endpoint_update_semaphore({ ... rest of the code  ... });
```

Reply:
Ah, you also want `with_endpoint_update_semaphore` to call shard 0. Makes sense

Reply:
Done in https://github.com/scylladb/scylladb/commit/67ce6376141a175411949ec355f2057d32ce00d4.